### PR TITLE
perf: implement cleanAndClose options for connections to reduce memory-footprint

### PR DESCRIPTION
**Pull request by OmkarHirve**
**Summary**

While working with mongoose, we faced this issue, where our memory keeps on increasing with the number of connections. While going through issues, found https://github.com/Automattic/mongoose/issues/2337. Also, saw a PR created but closed based on changes expected. https://github.com/Automattic/mongoose/pull/2338. Hence created this PR to address these changes.

**Examples**

`db.destroy();`

**NOTE** : PR is ready to merge.

### CODE REVIEW

**Comment by OmkarHirve** [comment]
@vkarpov15 I saw you reviewed the last PR. I request you to go through these changes and let me know if I am going in the correct direction or not. Also, any suggestion on this is most welcome. Thanks.

File: lib/connection.js
Comment (AbdelrahmanHafez) [inline lib/connection.js:931]: This condition wouldn't be matched if `closeAndClean` is undefined.
```suggestion
  if (closeAndClean != null) {
```

File: lib/connection.js
Comment (AbdelrahmanHafez) [inline lib/connection.js:960]: Please remove the experimental logs from the PR.

File: lib/connection.js
Comment (AbdelrahmanHafez) [inline lib/connection.js:984]: Same as above as well.

File: lib/connection.js
Comment (AbdelrahmanHafez) [inline lib/connection.js:996]: We need to remove this one as well.

**Review by AbdelrahmanHafez — CHANGES_REQUESTED** [review]
Welcome, @OmkarHirve 👋
Can you add some JS tests to prevent that functionality from being broken in the future, and add the necessary types to the respective `.d.ts` file?

Other than that, LGTM 👍

File: lib/connection.js
Comment (OmkarHirve) [reply lib/connection.js:960]: @AbdelrahmanHafez Yes, I am testing out these changes, will remove once testing is done, before I mark this ready to merge.

File: lib/connection.js
Comment (OmkarHirve) [reply lib/connection.js:984]: @AbdelrahmanHafez Yes, I am testing out these changes, will remove them once testing is done before I mark this ready to merge.

File: lib/connection.js
Comment (OmkarHirve) [reply lib/connection.js:996]: Yes, I am testing out these changes, will remove them once testing is done before I mark this ready to merge.

**Comment by OmkarHirve** [comment]
> Welcome, @OmkarHirve wave Can you add some JS tests to prevent that functionality from being broken in the future, and add the necessary types to the respective `.d.ts` file?
> 
> Other than that, LGTM +1

@AbdelrahmanHafez Sure, Thank you. Will do the change and update. Also, could you point any specific tests you would like me to run, so that all the changes are verified and are running as expeted?

**Comment by OmkarHirve** [comment]
Also, @AbdelrahmanHafez Change would be expected at this location as well, am I correct? https://github.com/Automattic/mongoose/blob/b56cd4a344ff91ae96c59bf6470d432b48b411ab/lib/drivers/node-mongodb-native/connection.js#L134

**Comment by vkarpov15** [comment]
A potential workaround: `mongoose.connections` is public. Why not just do `mongoose.connections.splice()` in your code to clean up dead connections that you do not intend to use anymore?

There's the same issue with this PR that was in #2338: you can still reopen a connection that was closed with `closeAndClean` set. To merge this, we'd need a check that throws an error if someone calls `Connection.prototype.openUri()` on a connection that's been already closed with `closeAndClean` set.

**Comment by OmkarHirve** [comment]
> A potential workaround: `mongoose.connections` is public. Why not just do `mongoose.connections.splice()` in your code to clean up dead connections that you do not intend to use anymore?
> 
> There's the same issue with this PR that was in #2338: you can still reopen a connection that was closed with `closeAndClean` set. To merge this, we'd need a check that throws an error if someone calls `Connection.prototype.openUri()` on a connection that's been already closed with `closeAndClean` set.

@vkarpov15 Thank you for your info on this thread.

1. As you mentioned using `mongoose.connections.splice()` in the particular code base, can be a workaround, True, but in case where people migrate their existing code base and start using mongoose, their time gets invested into looking why is the memory not getting cleaned up, hence I feel this change is necessary.
2. I will check on the point where it should throw an error if a cleaned connection is called to be reopened again.

**Comment by OmkarHirve** [comment]
> Welcome, @OmkarHirve wave Can you add some JS tests to prevent that functionality from being broken in the future, and add the necessary types to the respective `.d.ts` file?
> 
> Other than that, LGTM +1

Done. Please verify @AbdelrahmanHafez

**Comment by OmkarHirve** [comment]
> Also, @AbdelrahmanHafez Change would be expected at this location as well, am I correct?
> 
> https://github.com/Automattic/mongoose/blob/b56cd4a344ff91ae96c59bf6470d432b48b411ab/lib/drivers/node-mongodb-native/connection.js#L134
> 
> I am making the change, please suggest otherwise. Thanks

Found out that this is calling native mongo driver. No change is expected here.

**Comment by OmkarHirve** [comment]
@vkarpov15 @AbdelrahmanHafez I have added a unit test that verifies if the connection count gets reduced.

But as you mentioned @vkarpov15 we need to add a check that makes sure that if the cleaned connection is called up again for reconnection, it needs to throw an error. But I am not able to find out how could we actually reconnect with a closed connection. (Closed or not)

**Could you help me here please and provide an example unit test or some link where it already would have been used** ? This will make things quicker for me. 

Thanks

**Comment by OmkarHirve** [comment]
@Uzlopak @IslandRhythms @vkarpov15 @AbdelrahmanHafez Could you guys guide me here please? Thanks

**Comment by Uzlopak** [comment]
I dont know, but run `npm run lint -- --fix` to solve the linting issues

**Comment by OmkarHirve** [comment]
> @Uzlopak @IslandRhythms @vkarpov15 @AbdelrahmanHafez Could you guys guide me here please? Thanks

@Uzlopak @IslandRhythms @vkarpov15 @AbdelrahmanHafez 

I figured it out finally. Can someone re-run the workflows and review the PR, considering it is ready to merge?

**Comment by Uzlopak** [comment]
Hi, there are tests failing. Just ping me, to run the pipeline. I am basically addicted to github and would react fast.

**Comment by OmkarHirve** [comment]
@Uzlopak Sure I will,

I fixed the actual issue locally, but I am getting 
```
npm test

> mongoose@6.3.1 test /home/omkar/Documents/Sprints/mongoose
> mocha --exit ./test/connection.test.js



  1) "before all" hook in "{root}"

  0 passing (30s)
  1 failing

  1) "before all" hook in "{root}":
     Uncaught MongooseServerSelectionError: connect ECONNREFUSED 127.0.0.1:27017
      at NativeConnection.Connection.openUri (lib/connection.js:819:32)
      at Mongoose.createConnection (lib/index.js:286:10)
      at Module.module.exports (test/common.js:88:25)
      at dropDBs (test/common.js:163:21)
      at Context.<anonymous> (test/common.js:191:3)

npm ERR! Test failed.  See above for more details.
```
Do you have any idea on what type of failure is this?

**Comment by OmkarHirve** [comment]
@Uzlopak Thanks for helping me out on pipeline runs.

Could you also review this along with @vkarpov15 @AbdelrahmanHafez?

**Comment by OmkarHirve** [comment]
@vkarpov15 @AbdelrahmanHafez Any update?

File: lib/connection.js
Comment (Uzlopak) [inline lib/connection.js:920]: I would personally prefer to make the first parameter an options-object, with attributes force and closeAndClean. And if first parameter is of type boolean, the use it as a value for force. Currently it is backwards breaking and thus imho not acceptable.
Maybe also rename it. We call the method close so closeAndClean is doubled. So you can remove "closeAnd". On the other hand: what is meant with "clean"? Cleaning is here meant to be freeing of used resources. What is the resource? Sockets? Memory? Maybe should be called freeSockets or whatever it does.

Comment (Uzlopak) [reply lib/connection.js:920]: Ah wait, should be backwards compatible. But still the naming is strange.

Comment (OmkarHirve) [reply lib/connection.js:920]: @Uzlopak 

`force` is a type of string or an object and is sent to native-mongodb-driver, whereas `closeAndClean` has been introduced for cleaning up closed MongoDB connection. `cleanAndClose`  introduced with the close method should indicate that the user wants to clean that closed connection from memory and won't be used for re-connection.

Comment (Uzlopak) [reply lib/connection.js:920]: I just want to express my concern that "cleanAndClose" is a bad name. It should be destroyConnections or something like that to indicate the actual functionality.

Comment (OmkarHirve) [reply lib/connection.js:920]: Thank you @vkarpov15 @Uzlopak for your reviews on this PR.

I would be implementing the following changes. 

1. Instead of having a parameter in the `close` method, the user would call `conn.destroy()` 
2. conn.destroy()`  would take similar parameters to that of `close`, but in this method, the `destroyConnAfterClose` flag would be set to true.
3. Same would be checked in `openUri` method.

So from the user point of view, user would be calling `conn.destroy()` instead of `conn.close()`

@vkarpov15 @Uzlopak Please verify my understanding is correct, and I will be pushing in the changes by EOD today. 
Thanks

**Comment by OmkarHirve** [comment]
> I agree with @Uzlopak that the general functionality is good, but I'd prefer there be a separate `destroy()` method that closes the connection, ensures the connection will never be reopened, and removes it from the `connections` list.

@vkarpov15 Could you please verify my understanding:
https://github.com/Automattic/mongoose/pull/11674#discussion_r861434878

**Comment by Uzlopak** [comment]
First of all: Thank you for your work and being motivated.

@vkarpov15 is the maintainer. So it is his call.

**Comment by OmkarHirve** [comment]
@Uzlopak I would be requiring your help again regarding workflows. There is some issue in my local, hence I cannot run them locally. Could you trigger them please?

File: lib/connection.js
Comment (Uzlopak) [inline lib/connection.js:701]:
```suggestion
    const error = 'Connection has been closed and destroyed, and cannot be used for re-opening the connection. ' +
```

File: lib/connection.js
Comment (Uzlopak) [inline lib/connection.js:967]:
```suggestion
Connection.prototype._close = function(force, destroy, callback) {
```

Short and sexy :)

File: lib/connection.js
Comment (Uzlopak) [inline lib/connection.js:980]:
```suggestion
      if (destroyConnAfterClose && this.base.connections.indexOf(conn) !== -1) {
```
indexOf always returns a number, so we can use tripple equal :)

File: lib/connection.js
Comment (Uzlopak) [inline lib/connection.js:971]:
```suggestion
  this._destroyCalled = destroy;
```

File: lib/connection.js
Comment (Uzlopak) [inline lib/connection.js:974]:
```suggestion
    this.client._destroyCalled = destroy;
```

File: lib/connection.js
Comment (Uzlopak) [inline lib/connection.js:1002]:
```suggestion
        if (destroy && _this.base.connections.indexOf(conn) !== -1) {
```

File: lib/connection.js
Comment (Uzlopak) [inline lib/connection.js:1012]:
```suggestion
        _this.close(force, destroy, callback);
```

File: lib/connection.js
Comment (Uzlopak) [inline lib/connection.js:1018]:
```suggestion
        if (destroy && _this.base.connections.indexOf(conn) !== -1) {
```

File: test/connection.test.js
Comment (Uzlopak) [inline test/connection.test.js:399]:
```suggestion
        assert.equal(error.message, 'Connection has been closed and destroyed, and cannot be used for re-opening the connection. Please create a new connection with `mongoose.createConnection()` or `mongoose.connect()`.');
```

File: test/connection.test.js
Comment (Uzlopak) [inline test/connection.test.js:419]:
```suggestion
        assert.equal(error, 'Connection has been closed and destroyed, and cannot be used for re-opening the connection. Please create a new connection with `mongoose.createConnection()` or `mongoose.connect()`.');
```

**Review by Uzlopak — COMMENTED** [review]
Well, I did not expect that you did the changes without the feedback of vkarpov. I actually was thinking to just rename cleanAndClose to destroy and making destroy a short handle to close with  destroy always set to true.

Made PR remarks.

**Comment by OmkarHirve** [comment]
> Well, I did not expect that you did the changes without the feedback of vkarpov. I actually was thinking to just rename cleanAndClose to destroy and making destroy a short handle to close with destroy always set to true.
> 
> Made PR remarks.

@Uzlopak I made the changes based on changes suggested by @vkarpov15 in this comment.
https://github.com/Automattic/mongoose/pull/11674#discussion_r861211962

Also, thanks for the PR review, I will make the suggested changes.

**Comment by Uzlopak** [comment]
@OmkarHirve 
Do you think we should introduce a new state for the connection? Like STATES.destroyed?

**Comment by OmkarHirve** [comment]
@Uzlopak We currently store the state information in the connection object itself currently I believe, which would get removed if destroy is set to true. Once the connection object goes out of scope, I don't think then that would matter anymore? What is your thought on this?

**Comment by OmkarHirve** [comment]
@Uzlopak If you pull in my changes locally, could you check if unit tests run at your end? For me, unrelated tests are failing

File: lib/connection.js
Comment (Uzlopak) [inline lib/connection.js:700]: I mean here you check if _destroyCalled is set and ff the readyState is equal to disconnected. Would it not make sense to set the state to STATES.destroyed if we called with destroy?

Comment (OmkarHirve) [reply lib/connection.js:700]: Hmm, true @Uzlopak, till then the connection has been destroyed. This can be done I think. Currently, I will keep it as is, but @vkarpov15 could you provide information on if we can do this?

**Comment by Uzlopak** [comment]
@OmkarHirve 

I merged the rest of the suggestions and will check the failing tests.

File: lib/connection.js
Comment (Uzlopak) [inline lib/connection.js:963]:
```suggestion
 * @param {Boolean} destroy
```

File: lib/connection.js
Comment (Uzlopak) [inline lib/connection.js:980]:
```suggestion
      if (destroy && this.base.connections.indexOf(conn) !== -1) {
```

**Comment by Uzlopak** [comment]
Well. Based on your work I did this

https://github.com/Uzlopak/mongoose/tree/destroy-connection

**Comment by Uzlopak** [comment]
Are you ok with it that i create a PR based on your work?

**Comment by OmkarHirve** [comment]
> Are you ok with it that I create a PR based on your work?

Sorry @Uzlopak, I have put in much time and effort for these changes. I would be grateful if you rather not create a PR based on my changes. Also, that would encourage me to keep on contributing, if my contribution gets recognized.

**Comment by Uzlopak** [comment]
Sure mate, thats actually the reason why i asked :). Some people have just one PR and dont want to particilate more. Looking forward for your next work.

So you are obviously free to do the next steps.

File: test/connection.test.js
Comment (OmkarHirve) [reply test/connection.test.js]: @vkarpov15 
The test failures earlier forced me to try out some changes, and reverted those.

File: lib/connection.js
Comment (OmkarHirve) [reply lib/connection.js]: @vkarpov15 Done with suggested changes

**Comment by OmkarHirve** [comment]
@vkarpov15 Changes are done with all the mentioned suggestions. Could you please review and merge these changes?

**Comment by OmkarHirve** [comment]
@vkarpov15 Could you please review these changes?
thanks

File: test/connection.test.js
Comment (OmkarHirve) [reply test/connection.test.js:367]: Yes, we need, I am stubbing the call which is made to the actual MongoDB driver. Without it UTs were getting stuck, waiting on closure response.

Comment (Uzlopak) [reply test/connection.test.js:367]: When I wrote an alternative based on your PR, I did not need stubbing.

File: test/connection.test.js
Comment (Uzlopak) [reply test/connection.test.js:378]: For this you have the assert.equal in the next line which checks if the length is one smaller than the totalConn. You could rename totalConn to amountOfConnectionsBeforeClose.Also imho console log output like this is just unnecessary noise, when running the unit tests.

Comment (vkarpov15) [reply test/connection.test.js:378]: I agree about removing the `console.log()` statements. They're unnecessary given the `assert.equal()`. And while they might be handy if running this test in isolation, they're likely not helpful if running the whole test suite.

**Review by vkarpov15 — APPROVED** [review]
LGTM minus the `console.log()`. I'll also change the branch to v6.4 so we can ship this with our next minor release.

**Comment by OmkarHirve** [comment]
Thank you @vkarpov15 @Uzlopak.

**Comment by OmkarHirve** [comment]
@vkarpov15 Also, one question, does this not go to master? so that all versions after 6.4 would also have the changes?

**Comment by vkarpov15** [comment]
@OmkarHirve we'll merge 6.4 into master when we're ready to ship 6.4.0. We keep a separate branch so we can release patch releases while working on a minor release.

**Pull request merged**
